### PR TITLE
Add a unit to all fields

### PR DESCRIPTION
**Pull request by HealthyPear**
This PR adds a unit to all fields, in particular

- the default unit has been set to None
- new utility function to retrive subblock units as dictionaries
- new dtype builder for data subblocks
- added unit tests for all but run and event header, which depend on the version Corsika

This is both useful by itself and a precursor for #14.

### CODE REVIEW

File: corsikaio/subblocks/__init__.py [hunk 1]
```diff
@@ -13,11 +13,10 @@
 from ..constants import RUNH_VERSION_POSITION, EVTH_VERSION_POSITION
 
 __all__ = [
-    'parse_event_header',
-    'parse_run_header'
-    'parse_cherenkov_photons',
-    'parse_particle_data',
-    'parse_longitudinal',
+    "parse_event_header",
+    "parse_run_header" "parse_cherenkov_photons",
```
Comment (maxnoe) [inline corsikaio/subblocks/__init__.py:17]: Missing comma error that was here before now turns up here.

File: corsikaio/subblocks/run_header.py [hunk 2]
```diff
@@ -79,10 +79,17 @@
 
 
 def warn():
-    warnings.warn('Version unknown, using run header definition of version 7.x')
+    warnings.warn("Version unknown, using run header definition of version 7.x")
     return run_header_dtype_7x
 
 
+run_header_fields = defaultdict(warn)
```
Comment (maxnoe) [inline corsikaio/subblocks/run_header.py:86]: Warn returns the dtype, you need two functions now. One for the fields, one for the dtypes

File: corsikaio/subblocks/__init__.py [hunk 3]
```diff
@@ -63,3 +62,21 @@ def parse_particle_data(data_block_bytes):
...
+    subblock_fields: list(Field)
+        One of the defined lists of fields.
+
+    Return
+    ------
+    units: dict
+        Dictionary with field names as keys and
+        string representations of units as values.
+    """
+    units = {field.name: field.unit for field in subblock_fields}
```
Comment (maxnoe) [inline corsikaio/subblocks/__init__.py:81]: Not sure, but do we want to include the fields without units here? Maybe add `if field.unit is not None`?

File: corsikaio/subblocks/dtypes.py [hunk 4]
```diff
@@ -2,22 +2,27 @@
...
-            dt['formats'].append((field.dtype, field.shape))
+            dt["formats"].append((field.dtype, field.shape))
         else:
-            dt['formats'].append(field.dtype)
+            dt["formats"].append(field.dtype)
 
     return np.dtype(dict(**dt))
+
+
+def build_data_dtype(fields):
```
Comment (maxnoe) [inline corsikaio/subblocks/dtypes.py:26]: Why is that needed? Shouldn't the `build_dtype` function work for all lists of fields?

Comment (maxnoe) [reply corsikaio/subblocks/dtypes.py:26]: Which means I think, that `build_dtype(fields, itemsize=None)` shoudl work for that!

Comment (maxnoe) [reply corsikaio/subblocks/dtypes.py:26]: Yes, if you add do
```python
if itemsize is not None:
    dt["itemsize"] = itemsize
```

the second function is not needed.

File: corsikaio/subblocks/__init__.py (hunk 3 above)
Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py:81]: I assumed `units` needs to be as long as a possible `table.colnames` when one does,

`table = Table(data, units={field.name: field.unit for field in fields})`

If this is not true for me it's the same.

`unit` is defined anyway for each Field, fixing a default value makes sense to me

File: corsikaio/subblocks/dtypes.py (hunk 4 above)
Comment (HealthyPear) [reply corsikaio/subblocks/dtypes.py:26]: I did this because I notices that the resulting structured arrays didn't have also `offsets` and `formats` so I wanted to make sure I was replicating the dtype exactly.

should I also extend the condition to those 2 attributes?

File: corsikaio/subblocks/__init__.py (hunk 3 above)
Comment (maxnoe) [reply corsikaio/subblocks/__init__.py:81]: No, astropy goes through the dict and attaches units to the columns present in the dict. I think it's good for readability to leave out the `None` units.

File: corsikaio/subblocks/dtypes.py (hunk 4 above)
Comment (maxnoe) [reply corsikaio/subblocks/dtypes.py:26]: No, the defaults are fine for those.

File: corsikaio/subblocks/__init__.py (hunk 3 above)
Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py:81]: Will columns without units still be treated as dimensionless quantities?

Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py:81]: should we keep then the `None` default value in the specification of a new Field?

Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py:81]: Oh ok I checked it works anyway

Comment (maxnoe) [reply corsikaio/subblocks/__init__.py:81]: No, these will be `Columns` in a `QTable`. But there is no difference between putting `name: None` in the dict or not putting anything.

Comment (maxnoe) [reply corsikaio/subblocks/__init__.py:81]:
```
In [9]: t = QTable({"a": [1, 2, 3]}, units={"a": None})

In [10]: t
Out[10]: 
<QTable length=3>
  a  
int64
-----
    1
    2
    3

In [11]: t["a"]
Out[11]: 
<Column name='a' dtype='int64' length=3>
1
2
3

In [12]: t = QTable({"a": [1, 2, 3]}, units={})

In [13]: t
Out[13]: 
<QTable length=3>
  a  
int64
-----
    1
    2
    3

In [14]: t["a"]
Out[14]: 
<Column name='a' dtype='int64' length=3>
1
2
3

```

Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py:81]: mmmm I have 2 counter questions:

1. it feels weird to do `get_units_from_fields(some_subblock)` and not see all the original fields (especially if I don't know what am I looking at

2. wouldn't mixed column types mess up with some general operation on a table? if one needs to apply some row operation for example, he/she would have to specify `x.value` on Quantity columns, but not on normal Columns

Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py:81]: > there is no difference between putting name: None in the dict or not putting anything.

I don't see the same thing:

```python
import numpy as np
from astropy.table import QTable

units = {"b": "cm", "d": "ns"}
units = {"a": None, "b": "cm", "c": None, "d": "ns"}

a = np.array([1, 4, 5], dtype=np.int32)
b = [2.0, 5.0, 8.5]
c = ['x', 'y', 'z']
d = [10, 20, 30]

t = QTable([a, b, c, d], names=('a', 'b', 'c', 'd'), units=units)
t
```

This gives me a Quantity when specifying a dimensionless units, and a Column if not

Comment (maxnoe) [reply corsikaio/subblocks/__init__.py:81]: Regarding 1: that was my question... I am inclined that a function called `get_units...` only returns those fields that actually have units.

Regarding 2: that's an issue for astropy. I doubt they will change the `QTable` interface. You can easily check using `if table[col].unit is not None`:

Comment (maxnoe) [reply corsikaio/subblocks/__init__.py:81]: Runnig your example I get:

```
In [2]: t["a"]
Out[2]: 
<Column name='a' dtype='int32' length=3>
1
4
5

```

using astropy 5.1

File: corsikaio/subblocks/__init__.py (hunk 1 above)
Comment (HealthyPear) [reply corsikaio/subblocks/__init__.py]: fixed

File: corsikaio/subblocks/dtypes.py [hunk 5]
```diff
@@ -2,22 +2,24 @@
 import numpy as np
 
 
-Field = namedtuple('Field', ['position', 'name', 'shape', 'dtype'])
+Field = namedtuple("Field", ["position", "name", "unit", "shape", "dtype"])
 
-# set defaults for shape and dtype, this works for the 2 right most attributes
-Field.__new__.__defaults__ = (1, 'f4')
+# set defaults for shape and dtype, this works for the 3 right most attributes
```
Comment (maxnoe) [inline corsikaio/subblocks/dtypes.py:7]: unit, shape and dtype

File: corsikaio/subblocks/data.py [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
+    Field(1, "particle_description", dtype="float32"),
+    Field(2, "px", unit="Gev/c", dtype="float32"),
+    Field(3, "py", unit="Gev/c", dtype="float32"),
+    Field(4, "pz", unit="Gev/c", dtype="float32"),
+    Field(5, "x", unit="cm", dtype="float32"),
+    Field(6, "y", unit="cm", dtype="float32"),
+    Field(7, "t", unit="ns", dtype="float32"),
+]
+
+particle_data_dtype = build_dtype(particle_data_fields)
```
Comment (maxnoe) [inline corsikaio/subblocks/data.py:39]: You need to pass `itemsize=None` here, the default is `273 * 4`, which is the whole block.

File: corsikaio/subblocks/data.py [hunk 7]
```diff
@@ -1,33 +1,39 @@
...
+    Field(3, "y", unit="cm", dtype="float32"),
+    Field(4, "u", dtype="float32"),
+    Field(5, "v", dtype="float32"),
+    Field(6, "t", unit="ns", dtype="float32"),
+    Field(7, "wavelength", unit="nm", dtype="float32"),
+    Field(8, "mother_particle", dtype="uint16"),
+    Field(9, "production_height", unit="cm", dtype="float32"),
+]
+
+mmcs_cherenkov_photons_dtype = build_dtype(mmcs_cherenkov_photons_fields)
```
Comment (maxnoe) [inline corsikaio/subblocks/data.py:27]: You need to pass itemsize=None here, the default is 273 * 4, which is the whole block.

File: corsikaio/subblocks/data.py [hunk 8]
```diff
@@ -1,33 +1,39 @@
...
-    ('x', 'float32'),
-    ('y', 'float32'),
-    ('u', 'float32'),
-    ('v', 'float32'),
-    ('t', 'float32'),
-    ('wavelength', 'float32'),
-    ('mother_particle', 'uint16'),
-    ('production_height', 'float32'),
-])
+cherenkov_photons_dtype = build_dtype(cherenkov_photons_fields)
```
Comment (maxnoe) [inline corsikaio/subblocks/data.py:13]: You need to pass itemsize=None here, the default is 273 * 4, which is the whole block.

File: corsikaio/subblocks/data.py (hunk 7 above)
Comment (HealthyPear) [reply corsikaio/subblocks/data.py]: Fixed

File: corsikaio/subblocks/data.py (hunk 6 above)
Comment (HealthyPear) [reply corsikaio/subblocks/data.py]: Fixed

File: corsikaio/subblocks/dtypes.py (hunk 5 above)
Comment (HealthyPear) [reply corsikaio/subblocks/dtypes.py]: Fixed

File: corsikaio/subblocks/data.py (hunk 8 above)
Comment (HealthyPear) [reply corsikaio/subblocks/data.py]: Fixed

File: corsikaio/subblocks/data.py [hunk 9]
```diff
@@ -1,33 +1,39 @@
...
+    Field(8, "mother_particle", dtype="uint16"),
+    Field(9, "production_height", unit="cm", dtype="float32"),
+]
+
+mmcs_cherenkov_photons_dtype = build_dtype(mmcs_cherenkov_photons_fields, itemsize=None)
+
+particle_data_fields = [
+    Field(1, "particle_description", dtype="float32"),
+    Field(2, "px", unit="Gev/c", dtype="float32"),
+    Field(3, "py", unit="Gev/c", dtype="float32"),
```
Comment (maxnoe) [inline corsikaio/subblocks/data.py:32]: `Gev` → `GeV`  (everywhere)

**Review by maxnoe — COMMENTED** [review]
Minor comment, otherwise a very nice addition, thanks!

File: corsikaio/subblocks/data.py (hunk 9 above)
Comment (HealthyPear) [reply corsikaio/subblocks/data.py]: Fixed

File: tests/test_units.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    longitudinal_header_fields,
+)
+from corsikaio.subblocks.event_end import event_end_fields
+from corsikaio.subblocks.run_end import run_end_fields
+
+
+def test_new_field():
+
+    field = Field(1, "test_field")
+    assert field.unit == None
```
Comment (maxnoe) [inline tests/test_units.py:24]: comparison to `None` should be `is None`.

File: tests/test_units.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def test_event_end_units():
+
+    assert all([f.unit == None for f in event_end_fields])
```
Comment (maxnoe) [inline tests/test_units.py:29]: same

File: tests/test_units.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def test_run_end_units():
+
+    assert all([f.unit == None for f in run_end_fields])
```
Comment (maxnoe) [inline tests/test_units.py:34]: same

File: tests/test_units.py (hunk 10 above)
Comment (HealthyPear) [reply tests/test_units.py]: fixed

File: tests/test_units.py (hunk 11 above)
Comment (HealthyPear) [reply tests/test_units.py]: fixed

**Review by maxnoe — APPROVED** [review]
